internal/runtime/registry/registry.go

```diff
@@ -1,55 +1,206 @@
 	GroupVersionHook catalog.GroupVersionHook
-	ClientConfig     ClientConfig
+
+	ClientConfig   runtimev1.WebhookClientConfig
+	TimeoutSeconds *int32
+	FailurePolicy  *runtimev1.FailurePolicyType
+}
+
+type extensionRegistry struct {
+	ready bool
+	items map[string]*RuntimeExtensionRegistration
+	lock  sync.RWMutex
+}
+
+// Extensions provide access to all the registered Runtime Extensions.
+func Extensions() ExtensionRegistry {
+	once.Do(func() {
+		instance = extensions()
+	})
+	return instance
+}
+
+func extensions() ExtensionRegistry {
+	return &extensionRegistry{
+		items: map[string]*RuntimeExtensionRegistration{},
+	}
+}
+
+// WarmUp can be used to initialise a "cold" extension registry with all the known extensions at a given time.
+// After WarmUp completes the extension registry is considered ready.
+func (extensions *extensionRegistry) WarmUp(ext *runtimev1.ExtensionList) error {
+	if ext == nil {
+		return errors.New("Invalid argument, when calling WarmUp ext must not be nil")
+	}
+
+	extensions.lock.Lock()
+	defer extensions.lock.Unlock()
+
+	for i := range ext.Items {
+		if err := extensions.add(&ext.Items[i]); err != nil { // TODO: consider if to aggregate errors
+			return err
+		}
+	}
+
+	extensions.ready = true
+	return nil
 }
 
-type ClientConfig struct {
-	WebhookClientConfig runtimev1.WebhookClientConfig
-	TimeoutSeconds      *int32
-	FailurePolicy       *runtimev1.FailurePolicyType
-	NamespaceSelector   *metav1.LabelSelector
+// IsReady returns true if the extension registry is ready for usage, and this happens.
+func (extensions *extensionRegistry) IsReady() bool {
+	extensions.lock.RLock()
+	defer extensions.lock.RUnlock()
+
+	return extensions.ready
 }
 
-func (r registry) RegisterRuntimeExtension(ext *runtimev1.Extension) {
-	panic("implement me")
+// Add all the Runtime Extensions served by a registered runtime extension server.
+func (extensions *extensionRegistry) Add(ext *runtimev1.Extension) error {
+	if !extensions.ready {
+		return errors.New("Invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	if ext == nil {
+		return errors.New("Invalid argument, when calling Add ext must not be nil")
+	}
+
+	extensions.lock.Lock()
+	defer extensions.lock.Unlock()
+	return extensions.add(ext)
 }
 
-func (r registry) RemoveRuntimeExtension(ext *runtimev1.Extension) {
-	panic("implement me")
+// Remove all the Runtime Extensions served by a registered runtime extension server.
+func (extensions *extensionRegistry) Remove(ext *runtimev1.Extension) error {
+	if !extensions.ready {
+		return errors.New("Invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	if ext == nil {
+		return errors.New("Invalid argument, when calling Remove ext must not be nil")
+	}
+
+	extensions.lock.Lock()
+	defer extensions.lock.Unlock()
+
+	for _, e := range ext.Status.RuntimeExtensions { // TODO: consider if to delete using the extension name vs relying on status
```

Comment:
Not sure if you meant that but I think it would be safer to delete them by Extension.name (vs RuntimeExtensionRegistration.name).

Otherwise we depend on that the status still contains all the extensions we previously added.

Reply:
Related: Do we need an Update func too? (as we're running multiple Discoveries on the same Extension CR).

(or a combined Add/Update func)

Reply:
That's what I meant.
It is now fixed to delete by Registration name. Also Add acts already as Update if the RegistrationName already exists (I don't think we need a different method for this)

Reply:
Yup sounds good! One func is easier to use